ledger/complete/mtrie/node/node.go

```diff
@@ -175,24 +169,23 @@ func (n *Node) Compactify() *Node {
 		rChildEmpty = n.rChild.isDefaultNode()
 	}
 	if rChildEmpty && lChildEmpty {
-		// if both children are empty this is the same as as a default leafs
+		// if both children are empty this is the same as a default leaf
 		return nil
 	}
+	// If we reach the following code, at least one child must be a non-default node
 
-	// if childNode is non empty
+	// if childNode is non-empty
 	if rChildEmpty {
-		if !lChildEmpty && n.lChild.IsLeaf() {
+		if n.lChild.IsLeaf() {
 			return n.lChild.copyAndPromoteLeafNode(false)
 		}
-		// this would replace empty nodes with nil
-		n.rChild = nil
+		return NewInterimNode(n.height, n.lChild, nil)
```

Comment:
This change enforces immutability of `Node`, thereby increasing the safety of the code base at the expense of marginally more GC load. We will optimize the pruning to completely avoid unnecessary garbage generation in a subsequent PR.

Reply:
I totally agree this change enforces code safety and the code will change again _soon_, but I still think avoiding the extra GC cost is a simple change (return the same node, and make `Compactify` private) compared to the gain (I think the number of allocations from pruning is divided by 2). 
We would benefit from this gain till the new pruning logic is implemented.

Reply:
Its more than just a few lines, because `Compactify` is called by the Trie, which is a different package.  We can't just make `Compactify` private. Therefore, we need to introduce a new constructor for Nodes with compactification:
* I added the new constructor `NewInterimCompactifiedNode `  (-> [code](https://github.com/onflow/flow-go/blob/3afc1e5ded7bb158df49de01be852b29d9891e00/ledger/complete/mtrie/node/node.go#L121-L158)), and conslidated the code that was previously distributed over the `Compactify` and the `copyAndPromoteLeafNode`  methods. 
* The tests also needed refactoring, because we now construct the compactified node in one step and don't mutate a node anymore. 

To me, the code feels more concise now and is probably more performant.